c7n/output.py

```diff
@@ -323,6 +323,38 @@ def get_metadata(self):
         return res
 
 
+class JSONFormatter(logging.Formatter):
+    """
+    Convert Logs to JSON for more usable cloudwatch logs
+    https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/FilterAndPatternSyntax.html
+    """
+
+    def format(self, record):
+        json_log_object = {"type": "log",
+                           "logger": record.name,
+                           "log_time": datetime.fromtimestamp(record.created).isoformat(),
+                           "level": record.levelname,
+                           "module": record.module,
+                           "msg": self.convert_to_json(record.getMessage()),
+                           }
+
+        return json.dumps(json_log_object)
+
+    @staticmethod
+    def convert_to_json(message):
+        cnv_msg = {"original": message}
+        split_message = message.split()
+        for kv_pair in split_message:
+            if len(kv_pair.split(':')) == 2:
+                split = kv_pair.split(':')
+                try:
+                    val = ast.literal_eval(split[1])
```

Comment:
while this is cool technically, this feels quite silly. We're taking a bunch of parameters aka structured data, rendering it to a string, and then turning around again and trying to parse it back to structured data, thats a bunch of extra work. ast evaluation is also fairly slow to put in a data path like per log entry. I suggest instead the log message json format also capture keyword log messages in the record, and then punt on alternative forms then simple k:v, we'll address the individual log call sites as other changes are made.

Reply:
looks like its primarily positional only re what's in the log record but given `log.info("foo x:%s y:%z", x, y)` we can parse k:v on the uninterpolated msg and then pull the values from record.args.

Reply:
I dug a bit deeper on stdlib, I think there's a cleaner way to do this that will also get us to format strings usage and k:v pairs without the ambiguity or overhead on reparse, but that can be a followup layered on this.

Reply:
for now lets just yank on the ast parse and just support simple k:v pairs

Reply:
whats the timeline on moving to f strings since I think I saw cc support for python 2 for a lot of the year still.

Reply:
format strings in form literal f'' string, wasn't what I was referencing.. I was referencing using format strings of long form "k:{var_a} foo:{var_b}" in logs, but that will also give us a cleaner delineation of variables for json log output, in form of {'var_a': value, 'var_b': another_value} without resorting to string parsing.

Reply:
from reading the docs it seems that you would still have to do the string parsing since `log.info` , `log.debug` ultimately just returns a string so there is no way to pass the k:v pairs into the log object. it looks like you could pass in the k:v pairs in the kwarg `extra` but that would duplicate the data passed into the logger. https://docs.python.org/3/library/logging.html#logging.Logger.debug

happy to get your thoughts on this and more than willing to work on this in another pr if you want.

Reply:
its possible, but it requires changing the interface a little bit. I did discover there is a builtin in stdlib hack here that allows arg 1, to be. dict which also resolves.